iot-hub/Samples/device/DeviceReconnectionSample/DeviceReconnectionSample.cs

```diff
@@ -0,0 +1,227 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Azure.Devices.Client.Exceptions;
+using Microsoft.Extensions.Logging;
+using System;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Devices.Client.Samples
+{
+    public class DeviceReconnectionSample
+    {
+        private static readonly Random s_randomGenerator = new Random();
+        private const int TemperatureThreshold = 30;
+        private static readonly TimeSpan s_sleepDuration = TimeSpan.FromSeconds(5);
+        private static readonly TimeSpan s_operationTimeout = TimeSpan.FromHours(1);
+
+        private readonly string _deviceConnectionString;
+        private readonly TransportType _transportType;
+        private readonly ILogger _logger;
+        private readonly object _lock = new object();
+
+        private static DeviceClient s_deviceClient;
+        private static ConnectionStatus connectionStatus;
+        private static bool wasConnected;
+
+        public DeviceReconnectionSample(string deviceConnectionString, TransportType transportType, ILogger logger)
+        {
+            _logger = logger;
+            _deviceConnectionString = deviceConnectionString;
+            _transportType = transportType;
+
+            InitializeClient();
+        }
+
+        private void ConnectionStatusChangeHandler(ConnectionStatus status, ConnectionStatusChangeReason reason)
+        {
+            _logger.LogDebug($"Connection status changed: status={status}, reason={reason}");
+
+            connectionStatus = status;
+            switch (connectionStatus)
+            {
+                case ConnectionStatus.Connected:
+                    _logger.LogDebug("### The DeviceClient is CONNECTED; all operations will be carried out as normal.");
+
+                    wasConnected = true;
+                    break;
+
+                case ConnectionStatus.Disconnected_Retrying:
+                    _logger.LogDebug("### The DeviceClient is retrying based on the retry policy. Do NOT close or open the DeviceClient instance");
```

Comment:
shouldn't wasConnected be set to false?

Reply:
I am setting it inside the block where dispose is being called (after the expression using this variable has been evaluated).

Reply:
Probably stemmed from my misunderstanding on what `wasConnected` meant. It should probably be initialized to false and then only changed to true once, right?

Reply:
yes, since it is a bool, it defaults to `false`. It is set to true when the client reports `ConnectionStatus.Connected`. After the client is disposed by the calling application, I reset it to false, so that the subsequent initialization and reconnection will again start fresh.

Reply:
makes sense